File: backend/app/services/invoice_service.py
from decimal import Decimal

from app.schemas.invoice import InvoiceItemCreate


def calculate_invoice_totals(items: list[InvoiceItemCreate]) -> dict:
    subtotal = Decimal("0.00")
    tax_total = Decimal("0.00")
    discount_total = Decimal("0.00")
    parsed_items: list[dict] = []

    for item in items:
        line_subtotal = item.quantity * item.unit_price
        line_discount = item.discount_amount
        line_taxable = line_subtotal - line_discount
        line_tax = line_taxable * (item.tax_rate / Decimal("100.00"))
        line_total = line_taxable + line_tax

        subtotal += line_subtotal
        tax_total += line_tax
        discount_total += line_discount

        parsed_items.append(
            {
                "description": item.description,
                "quantity": item.quantity,
                "unit_price": item.unit_price,
                "tax_rate": item.tax_rate,
                "discount_amount": item.discount_amount,
                "line_total": line_total,
            }
        )

    total_amount = subtotal - discount_total + tax_total
    return {
        "subtotal": subtotal,
        "tax_amount": tax_total,
        "discount_amount": discount_total,
        "total_amount": total_amount,
        "balance_due": total_amount,
        "items": parsed_items,
    }

File: frontend/lecoquette/src/services/invoice.service.ts
import { apiRequest } from "./api";
import { authService } from "./auth.service";

export interface InvoiceItem {
  id?: number;
  description: string;
  quantity: number;
  unit_price: number;
  tax_rate: number;
  discount_amount: number;
  line_total?: number;
}

export type InvoiceStatus = "draft" | "paid" | "overdue" | "cancelled";

export interface Invoice {
  id: number;
  customer_id: number;
  invoice_number?: string;
  customer?: string;
  cliente?: string;
  issue_date?: string;
  fecha?: string;
  due_date?: string;
  fechaVencimiento?: string;
  status?: string;
  estado?: string;
  currency_code?: string;
  subtotal?: number;
  tax_amount?: number;
  discount_amount?: number;
  total_amount?: number;
  total?: number;
  monto?: number;
  balance_due?: number;
  notes?: string;
  created_at?: string;
  updated_at?: string;
  items?: InvoiceItem[];
}

export interface CreateInvoicePayload {
  customer_id: number;
  issue_date: string;
  due_date: string;
  currency_code: string;
  notes: string;
  status: string;
  items: {
    description: string;
    quantity: number;
    unit_price: number;
    tax_rate: number;
    discount_amount: number;
  }[];
}

export const invoiceService = {
  async getAll(): Promise<Invoice[]> {
    return apiRequest<Invoice[]>("/invoices/", {
      token: authService.getToken(),
    });
  },

  async create(payload: CreateInvoicePayload): Promise<Invoice> {
    return apiRequest<Invoice>("/invoices/", {
      method: "POST",
      body: payload,
      token: authService.getToken(),
    });
  },

  async updateStatus(invoiceId: number, status: InvoiceStatus): Promise<Invoice> {
    return apiRequest<Invoice>(`/invoices/${invoiceId}/status`, {
      method: "PUT",
      body: { status },
      token: authService.getToken(),
    });
  },

  async remove(invoiceId: number): Promise<void> {
    return apiRequest<void>(`/invoices/${invoiceId}`, {
      method: "DELETE",
      token: authService.getToken(),
    });
  },
};